client/dash/client_wallet.go

```
@@ -0,0 +1,482 @@
+	}
+	address, err := btcutil.DecodeAddress(addr, w.Params())
+	if err != nil {
+		return "", err
+	}
+	return w.GetPrivKeyForAddress(pw, address)
+}
+
+func (ec *DashElectrumClient) SignTx(pw string, txBytes []byte) ([]byte, error) {
+	w := ec.GetWallet()
+	if w == nil {
+		return nil, ErrNoWallet
+	}
+	w.UpdateTip(ec.Tip())
+	// Note: this errors if no inputs or no outputs or both
+	unsignedTx, err := newWireTx(txBytes, true /*checkIo*/)
+	if err != nil {
+		return nil, err
+	}
+	var signInfo = &wallet.SigningInfo{
+		UnsignedTx: unsignedTx,
+		VerifyTx:   true,
+	}
+	return w.SignTx(pw, signInfo)
+}
+
+// GetWalletTx gets a tx from this wallet if it exists. If it does not esist then
+// we return error.
+// Edge cases:
+//
+//   - If the tx in the db is unmined we return it with no error. But if it has been
+//     mined then we compare the height it was mined with 'maybeTip' and return an
+//     error if the wallet tip is obviously behind electrumx; and we also return a
+//     flag indicating that.
+//     Software calling this API can then choose to directly ask electrumx for tx
+//     information using the GetTransaction API.
+//
+//   - if wallet tx height is less than or equal to 'maybeTip' we return the tx with
+//     no error even though we may be behind electrumx.
+func (ec *DashElectrumClient) GetWalletTx(txid string) (int, bool, []byte, error) {
+	w := ec.GetWallet()
+	if w == nil {
+		return -1, false, nil, ErrNoWallet
+	}
+	txn, err := w.GetTransaction(txid)
+	if err != nil {
+		// 'no such transaction'
+		return -1, false, nil, err
+	}
+	if txn.Height < 0 {
+		return -1, false, nil, errors.New("txns db error")
+	}
+	// not mined yet? return valid tx with no confs
+	if txn.Height == 0 {
+		return 0, false, txn.Bytes, nil
+	}
+	// mined
+	maybeTip := ec.Tip()
+	conf := maybeTip - txn.Height
+	// definitely client tip is behind electrumx
+	if conf < 0 {
+		return int(conf), true, txn.Bytes, errors.New("tip behind electrumx")
+	}
```

Odd this returns an error and info. Isn't the bool already used to determine this case? I don't think this should return error.

---

I remember this as tricky bit of logic when I did a test integration with Dex maybe a year ago and I was calling from dex code .. I think it should return an error for sure but the bool I cannot remember too well .. I think it was to indicate that the transaction had been mined and was a 'don't care' return if there was an error also.

Note: The wallet implementation for BTC is written and tested and works. The Firo & Dash wallets are placeholders containing the BTC wallet code and their respective client implementations are changed enough to build and run the back ends which are truly tested live over an extended period.

I do not want to change this until I can run the BTC implementation live with some test integration code .. And this may all be moot depending on @buck54321 intentions re: `bisoncraft/utxowallet.